android/modules/ui/src/java/ti/modules/titanium/ui/widget/TiUILabel.java

```diff
@@ -88,99 +99,176 @@ protected void onLayout(boolean changed, int left, int top, int right, int botto
-						} else {
-							tv.setTextSize(TiUIHelper.getSizeUnits(null), TiUIHelper.getSize(null));
-						}
-					}
-				}
-			    
-			    TextPaint textPaint = tv.getPaint();
-				if (textPaint != null) {
-					float stringWidth = textPaint.measureText(tv.getText().toString());
-					int textViewWidth = tv.getWidth();
-					if (textViewWidth < stringWidth && stringWidth > 0) {
-						float fontSize = (textViewWidth / stringWidth) * tv.getTextSize();
-						autoshrinkSetFontSize = fontSize > TiConvert.toFloat(minimumFontSize, 0) ? String.valueOf(fontSize) : minimumFontSize;
-						tv.setTextSize(TiUIHelper.getSizeUnits(autoshrinkSetFontSize), TiUIHelper.getSize(autoshrinkSetFontSize));
-					}
-				}
+	private void adjustTextFontSize(View view)
+	{
+		// Do not continue if auto-sizing the font is disabled.
+		if (this.minimumFontSizeInPixels < FONT_SIZE_EPSILON) {
+			return;
+		}
+
+		// Make sure the given view is a TextView.
+		if ((view instanceof TextView) == false) {
+			return;
+		}
+		TextView textView = (TextView)view;
+
+		// Fetch the display's density scale and calculate pixel's per virtual point, rounded up.
+		// Note: To keep it simple, don't let the scale be below 1x.
+		//       Only obsolete "ldpi" devices use scales lower than this.
+		float densityScale = view.getResources().getDisplayMetrics().density;
+		if (densityScale <= 0) {
+			densityScale = 1.0f;
+		}
+
+		// Calculate the pixel width within the view that the text will be rendered to.
+		float viewContentWidth;
+		{
+			int value = textView.getWidth();
+			{
+				// Exlude the view's padding and borders.
+				value -= textView.getTotalPaddingLeft() + textView.getTotalPaddingRight();
+			}
+			if ((this.layoutParams != null) &&
+			    (this.layoutParams.optionWidth != null) && !layoutParams.autoFillsWidth)
+			{
+				// Shave off 1 dp (device-independent pixel) from the width.
+				// This prevents ellipsis from being applied for text that just-fits within the view.
+				// Note: We don't want to do this if the view is auto-sized to fit its text
+				//       because the font/text might shrink on every call to onLayout() method.
+				int pixelsPerPoint = (int)Math.ceil((double)densityScale);	// Round up.
+				value -= pixelsPerPoint;
+			}
+			if (value <= 0) {
+				return;
+			}
+			viewContentWidth = (float)value;
+		}
+
+		// Fetch the view's text.
+		String text = null;
+		if (textView.getText() != null) {
+			text = textView.getText().toString();
+		}
+		if ((text == null) || (text.length() <= 0)) {
+			return;
+		}
+
+		// Restore the view's original font size.
+		float previousFontSize = textView.getTextSize();
+		textView.setTextSize(android.util.TypedValue.COMPLEX_UNIT_PX, this.unscaledFontSizeInPixels);
+
+		// Automatically downscale the font size to fit the width of the view.
+		// Note: Downscaling text width-wise is inaccurate and may take a few attempts.
```

Comment:
Unfortunately no.

The issue here is that non-monospaced fonts have variable character width. For example, the characters 'i' and 'w' have radically different widths. There are also character ligatures such as 'Æ'. And to make things even more complicated are cursive language, such as Arabic, where the text renderer will combine characters, making per character pixel width measurement kind of useless.

The issue the old code had before is that it would not always scale the font size down enough, causing text to be clipped. Such as with a string containing many wide characters.